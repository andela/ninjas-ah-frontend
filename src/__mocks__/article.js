export default (Object.article = {
  title: 'my title well done',
  description: 'description of the article: well',
  body: 'body of the article: well'
});
