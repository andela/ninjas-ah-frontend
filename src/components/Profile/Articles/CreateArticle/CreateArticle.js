import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import MetaTags from 'react-meta-tags';
import { EditorState, convertToRaw } from 'draft-js';
import { Editor } from 'react-draft-wysiwyg';
import 'react-draft-wysiwyg/dist/react-draft-wysiwyg.css';
import TextareaAutosize from 'react-autosize-textarea';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faQuestionCircle } from '@fortawesome/free-solid-svg-icons';
import richTextConfig from '../../../../helpers/richTextConfig.json';
import { Alert } from '../../../common';
import { createPost } from '../../../../actions';
import '../EditArticle/EditArticle.scss';
import Layout from '../../../Layout';

export class CreateArticle extends Component {
  state = {
    page: 'Create a story',
    title: '',
    description: '',
    body: '',
    coverUrl: null,
    article: '',
    errors: {},
    editorState: EditorState.createEmpty()
  };

  componentWillReceiveProps(nextProps) {
    const { history } = this.props;
    if (nextProps.errors) {
      this.setState({ errors: nextProps.errors });
    }
    if (nextProps.article && Object.keys(nextProps.article).length) {
      this.setState({ article: nextProps.article });
      history.push(`/profile/article/preview/${nextProps.article.slug}`);
    }
  }

  onChange = (e) => {
    this.setState({ errors: [], [e.target.name]: e.target.value });
  };

  onEditorStateChange = (editorState) => {
    this.setState({ editorState });
  };

  fileSelectedHandler = (e) => {
    this.setState({ coverUrl: e.target.files[0] });
  };

  onSubmit = (e) => {
    e.preventDefault();
    const formData = new FormData();
    const body = this.state.editorState
      ? convertToRaw(this.state.editorState.getCurrentContent())
      : '';
    formData.append('title', this.state.title);
    formData.append('description', this.state.description);
    formData.append('body', JSON.stringify(body));
    formData.append('coverUrl', this.state.coverUrl);
    this.props.createPost(formData);
  };

  render() {
    const { page, message, errors, article, editorState } = this.state;

    return (
      <Layout>
        <div id="createArticle">
          <MetaTags>
            <title>Create Article - Authors Haven</title>
            <meta
              name="description"
              content="Create a community of like minded authors to foster inspiration and innovation by leveraging the modern web"
            />
          </MetaTags>
          <div className="container">
            <div className="row">
              {Object.entries(errors || []).map((value, key) => (
                <div className="text-danger border b-light medium-padding text-white" key={key}>
                  <FontAwesomeIcon icon={faQuestionCircle} /> {value[1]}
                </div>
              ))}
              {article.title ? (
                <Alert
                  loading={true}
                  alertType="success"
                  message="Article created successfully. We are redirecting you..."
                />
              ) : (
                ''
              )}
              {message ? <Alert alertType="success" message={message} /> : ''}
              <form id="saveArticle" onSubmit={this.onSubmit}>
                {message ? <Alert alertType="success" message={message} /> : ''}
                <div className="row">
                  <h2 className="large-v-padding">{page}</h2>
                  <div className="input-field">
                    <label>Title</label>
                    <br />
                    <TextareaAutosize
                      name="title"
                      onChange={this.onChange}
                      value={this.state.title}
                      placeholder="Type title here..."
                      style={{ minHeight: 20 }}
                      id="articleTitle"
                    />
                  </div>
                  <div className="input-field">
                    <label>Description</label>
                    <br />
                    <TextareaAutosize
                      name="description"
                      onChange={this.onChange}
                      value={this.state.description}
                      placeholder="Type description here..."
                      style={{ minHeight: 20 }}
                      id="articleDescription"
                    />
                  </div>
                  <div className="input-field">
                    <label>Upload Cover Image</label>
                    <input
                      type="file"
                      name="coverUrl"
                      value={this.coverUrl}
                      onChange={this.fileSelectedHandler}
                    />
                  </div>
                  <div className="input-field">
                    <label>Body of the article</label>
                    <br />
                    <Editor
                      name="body"
                      editorState={editorState}
                      toolbarClassName="toolbarClassName"
                      wrapperClassName="wrapperClassName"
                      editorClassName="editorClassName"
                      value={this.state.body}
                      placeholder="Body of the article..."
                      onEditorStateChange={this.onEditorStateChange}
                      toolbar={richTextConfig}
                      style={{ minHeight: 15 }}
                      id="articleBody"
                    />
                  </div>
                </div>
                <button type="submit" className="button radius-4 primary text-black bold">
                  Save
                </button>
              </form>
            </div>
          </div>
        </div>
      </Layout>
    );
  }
}
CreateArticle.propTypes = {
  createPost: PropTypes.func.isRequired,
  article: PropTypes.object,
  errors: PropTypes.any,
  history: PropTypes.object
};
const mapStateToProps = ({ articles: { article, errors } }) => ({
  article,
  errors
});

export default connect(
  mapStateToProps,
  { createPost }
)(CreateArticle);
