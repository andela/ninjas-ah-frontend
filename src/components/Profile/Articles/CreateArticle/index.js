import CreateArticle from './CreateArticle';

export default CreateArticle;
