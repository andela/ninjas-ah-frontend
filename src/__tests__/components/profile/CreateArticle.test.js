import React from 'react';
import configureMockStore from 'redux-mock-store';
import thunk from 'react-thunk';
import { Provider } from 'react-redux';
import article from '../../../__mocks__/article';
import CreateArticle, { CreateArticle as CreateArticleComponent } from '../../../components/Profile/Articles/CreateArticle/CreateArticle';
import { shallow, mount } from '../../../../config/enzymeConfig';

describe('<CreateArticle />', () => {
  const props = {
    errors: { error: ['article not found'] },
    article: {
      title: 'Hello John Doe',
      description: 'John Doe, Mocker',
      body: 'body of the article'
    },
    onCreateArticle: jest.fn(),
    history: { push: jest.fn() },
    onEditorStateChange: jest.fn(),
    createPost: jest.fn()
  };

  const component = shallow(<CreateArticleComponent {...props} />);
  it('should render a <CreateArticleComponent /> component ', () => {
    expect(component).toMatchSnapshot();
  });
  it('should call onChange method when the title value is changed', () => {
    const spy = jest.spyOn(component.instance(), 'onChange');
    component.instance().forceUpdate();
    const event = { target: { value: 'title' } };
    const textArea = component.find('#articleTitle');
    textArea.simulate('change', event);
    expect(textArea.length).toBe(1);
    expect(spy).toHaveBeenCalled();
  });
  it('should call onChange method when the description value is changed', () => {
    const spy = jest.spyOn(component.instance(), 'onChange');
    component.instance().forceUpdate();
    const event = { target: { value: 'description' } };
    const textArea = component.find('#articleDescription');
    textArea.simulate('change', event);
    expect(textArea.length).toBe(1);
    expect(spy).toHaveBeenCalled();
  });
  it('should call onChange method when the body value is changed', () => {
    const instance = component.instance();
    instance.onEditorStateChange();
    expect(instance).toBeDefined();
  });
  it('should onFileChange', () => {
    const instance = component.instance();
    const fakeEvent = { target: { files: ['name'] } };
    instance.fileSelectedHandler(fakeEvent);
    expect(instance).toBeDefined();
  });
  it('should create articles', () => {
    const instance = component.instance();
    const fakeEvent = { preventDefault: () => {} };
    instance.onSubmit(fakeEvent);
  });
  it('should create articles', () => {
    const instance = component.instance();
    const fakeEvent = { preventDefault: () => {} };
    component.setProps({ article });
    instance.onSubmit(fakeEvent);
  });
});
