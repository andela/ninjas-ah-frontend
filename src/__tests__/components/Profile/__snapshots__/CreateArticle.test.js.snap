// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateArticle /> should render a <CreateArticleComponent /> component  1`] = `ShallowWrapper {}`;
